wolfcrypt/src/port/Renesas/renesas_rx64_hw_sha.c

```diff
@@ -0,0 +1,424 @@
+        } else if(sha_type == RX64_SHA224) {
+            ret = R_Sha224_HashDigest(&data[index], out, chunk_length, flag, &work_sha.sha224);
+        } else if(sha_type == RX64_SHA256) {
+            ret = R_Sha256_HashDigest(&data[index], out, chunk_length, flag, &work_sha.sha256);
+        }
+        if (ret != R_PROCESS_COMPLETE) {
+            // On failure break, unlock hardware, return error.
+            break;
+        }
+        index += chunk_length;
+        flag = R_SHA_ADD;
+    } while (index < len);
+
+    rx64_hw_unlock();
+    return ret;
+}
+
+/**
+ * @brief Free a hash for use with the RX64 SHA subsystem.
+ *
+ * @param[in] hash The hash to free
+ */
+static void RX64_HashFree(wolfssl_RX64_HW_Hash* hash)
+{
+    if (hash == NULL)
+        return;
+
+    if (hash->msg != NULL) {
+        XFREE(hash->msg, hash->heap, DYNAMIC_TYPE_TMP_BUFFER);
+        hash->msg = NULL;
+    }
+}
+
+/**
+ * @brief Initialize a hash for use with the RX64 SHA subsystem.
+ *
+ * @param[in] hash The hash to initialize
+ * @param[in] heap Optional pointer to memory to use.
+ * @param devId Unused
+ * @param[in] sha_type The SHA type for this hash
+ * @return int 0 on success, BAD_FUNC_ARG on failure
+ */
+static int RX64_HashInit(wolfssl_RX64_HW_Hash* hash, void* heap, int devId,
+    word32 sha_type)
+{
+    if (hash == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    (void)devId;
+    XMEMSET(hash, 0, sizeof(wolfssl_RX64_HW_Hash));
+
+    hash->heap = heap;
+    hash->len  = 0;
+    hash->used = 0;
+    hash->msg  = NULL;
+    hash->sha_type = sha_type;
+
+    return 0;
+}
+
+/**
+ * @brief Add data to the hash with the RX64 SHA subsystem.
+ *
+ * Note that do to the limitations in the RX64 hardware
+ * and it's inability to save the current state,
+ * this function actually just adds the data to a buffer
+ * that will then be processed when calling HashFinal or HashGet
+ *
+ * @param[in] hash Hash structure
+ * @param[in] data data to hash
+ * @param[in] sz size of the data
+ * @return int 0 on success, BAD_FUNC_ARG or MEMORY_E on failure
+ */
+static int RX64_HashUpdate(wolfssl_RX64_HW_Hash* hash, const byte* data, word32 sz)
+{
+    if (hash == NULL || (sz > 0 && data == NULL)) {
+        return BAD_FUNC_ARG;
+    }
+
+    if (hash->len < hash->used + sz) {
+        if (hash->msg == NULL) {
+            hash->msg = (byte*)XMALLOC(hash->used + sz, hash->heap,
+                    DYNAMIC_TYPE_TMP_BUFFER);
+        } else {
+            byte* pt = (byte*)XREALLOC(hash->msg, hash->used + sz, hash->heap,
+                    DYNAMIC_TYPE_TMP_BUFFER);
+            if (pt == NULL) {
+                return MEMORY_E;
+            }
+            hash->msg = pt;
+        }
+        if (hash->msg == NULL) {
+            return MEMORY_E;
+        }
+        hash->len = hash->used + sz;
+    }
+    XMEMCPY(hash->msg + hash->used, data , sz);
+    hash->used += sz;
+
+    return 0;
+}
```

Comment:
Because TSIP and SCE don't allow multiple message digest process in parallel, TSIP and SCE implementation stores message data into internal buffer so that calculating sha message is done by one call. RX64 and RX71 have the same limitation?

Reply:
Yes the RX64/RX71 have the same limitation